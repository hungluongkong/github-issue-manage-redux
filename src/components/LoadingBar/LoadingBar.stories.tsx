import React from 'react';
import LoadingBar from '.';
import { storiesOf } from '@storybook/react';
import { PreviewArea, Code } from '../commonStyled';

storiesOf('Loading Bar', module)
  .add('Default', (): JSX.Element => (
    <>
      <h3>Loading Bar</h3>
      <PreviewArea>
        <LoadingBar />
      </PreviewArea>
      <Code>
        {`
          <LoadingBar />
        `}
      </Code>
    </>
  ));
