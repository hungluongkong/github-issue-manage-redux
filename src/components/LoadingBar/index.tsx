import * as React from 'react';

import styled, { keyframes } from 'styled-components';
import * as color from '../../theme/color';

const loadingKeyFrame = keyframes`
  0% {
    top: 8px;
    height: 64px;
  }
  50%, 100% {
    top: 24px;
    height: 32px;
  }
`;

const LoadingWrapper = styled.div`
  text-align: center;
`;

const LoadBar = styled.div`
  display: inline-block;
  position: relative;
  width: 80px;
  height: 80px;
`;

const LoadingBlock = styled.div`
  display: inline-block;
  position: absolute;
  left: 8px;
  width: 16px;
  background: ${color.LOADING_BAR_COLOR};
  animation: ${loadingKeyFrame} 1.2s cubic-bezier(0, 0.5, 0.5, 1) infinite;

  :nth-child(1) {
    left: 8px;
    animation-delay: -0.24s;
  }
  :nth-child(2) {
    left: 32px;
    animation-delay: -0.12s;
  }
  :nth-child(3) {
    left: 56px;
    animation-delay: 0;
  }
`;

const LoadingBar = (): JSX.Element => (
  <LoadingWrapper>
    <LoadBar>
      {/* Three loading box effect */}
      <LoadingBlock />
      <LoadingBlock />
      <LoadingBlock />
    </LoadBar>
  </LoadingWrapper>
);

export default LoadingBar;
