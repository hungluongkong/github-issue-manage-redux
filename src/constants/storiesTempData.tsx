/**
 * Template response json string from github.
 * This is a list of issues.
 * URL: https://api.github.com/repos/hungluong-asnet/demo-api/issues
 */
export const issueJson: string = `
{
  "url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1",
  "repository_url": "https://api.github.com/repos/hungluong-asnet/demo-api",
  "labels_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1/labels{/name}",
  "comments_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1/comments",
  "events_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1/events",
  "html_url": "https://github.com/hungluong-asnet/demo-api/issues/1",
  "id": 653740906,
  "node_id": "MDU6SXNzdWU2NTM3NDA5MDY=",
  "number": 1,
  "title": "isssue no 1",
  "user": {
    "login": "hungluong-asnet",
    "id": 68043715,
    "node_id": "MDQ6VXNlcjY4MDQzNzE1",
    "avatar_url": "https://avatars1.githubusercontent.com/u/68043715?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/hungluong-asnet",
    "html_url": "https://github.com/hungluong-asnet",
    "followers_url": "https://api.github.com/users/hungluong-asnet/followers",
    "following_url": "https://api.github.com/users/hungluong-asnet/following{/other_user}",
    "gists_url": "https://api.github.com/users/hungluong-asnet/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/hungluong-asnet/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/hungluong-asnet/subscriptions",
    "organizations_url": "https://api.github.com/users/hungluong-asnet/orgs",
    "repos_url": "https://api.github.com/users/hungluong-asnet/repos",
    "events_url": "https://api.github.com/users/hungluong-asnet/events{/privacy}",
    "received_events_url": "https://api.github.com/users/hungluong-asnet/received_events",
    "type": "User",
    "site_admin": false
  },
  "labels": [

  ],
  "state": "open",
  "locked": false,
  "assignee": null,
  "assignees": [

  ],
  "milestone": null,
  "comments": 0,
  "created_at": "2020-07-09T03:48:36Z",
  "updated_at": "2020-07-09T03:48:36Z",
  "closed_at": null,
  "author_association": "OWNER",
  "active_lock_reason": null,
  "body": "Nothing inside issue"
}
`;

export const issueListJson: string = `
  [
    {
      "url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2",
      "repository_url": "https://api.github.com/repos/hungluong-asnet/demo-api",
      "labels_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2/labels{/name}",
      "comments_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2/comments",
      "events_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2/events",
      "html_url": "https://github.com/hungluong-asnet/demo-api/issues/2",
      "id": 654622112,
      "node_id": "MDU6SXNzdWU2NTQ2MjIxMTI=",
      "number": 2,
      "title": "#practice #make #api",
      "user": {
        "login": "hungluong-asnet",
        "id": 68043715,
        "node_id": "MDQ6VXNlcjY4MDQzNzE1",
        "avatar_url": "https://avatars1.githubusercontent.com/u/68043715?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/hungluong-asnet",
        "html_url": "https://github.com/hungluong-asnet",
        "followers_url": "https://api.github.com/users/hungluong-asnet/followers",
        "following_url": "https://api.github.com/users/hungluong-asnet/following{/other_user}",
        "gists_url": "https://api.github.com/users/hungluong-asnet/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/hungluong-asnet/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/hungluong-asnet/subscriptions",
        "organizations_url": "https://api.github.com/users/hungluong-asnet/orgs",
        "repos_url": "https://api.github.com/users/hungluong-asnet/repos",
        "events_url": "https://api.github.com/users/hungluong-asnet/events{/privacy}",
        "received_events_url": "https://api.github.com/users/hungluong-asnet/received_events",
        "type": "User",
        "site_admin": false
      },
      "labels": [

      ],
      "state": "open",
      "locked": false,
      "assignee": null,
      "assignees": [

      ],
      "milestone": null,
      "comments": 0,
      "created_at": "2020-07-10T08:47:16Z",
      "updated_at": "2020-07-10T08:47:16Z",
      "closed_at": null,
      "author_association": "OWNER",
      "active_lock_reason": null,
      "body": "**Don't know why this issue is here. and why we need to create it?**"
    },
    ${issueJson}
  ]`;
