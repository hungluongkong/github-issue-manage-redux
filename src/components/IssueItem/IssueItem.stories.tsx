import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import IssueItem from '.';
import { JsonToIssueList } from '../../helpers/jsonHelper';
import { issueListJson } from '../../constants/storiesTempData';
import { PreviewArea, Code, DetailTable } from '../commonStyled';

// Detail of props
const propsDetail: JSX.Element = DetailTable({
  data: [
    { name: 'issue', type: 'interface Issue', description: 'Issue to display' },
  ],
});

const [issue] = JsonToIssueList(issueListJson);

storiesOf('Issue Item', module).add(
  'Issue',
  (): JSX.Element => (
    <>
      <h3>Default Item</h3>
      <PreviewArea>
        <IssueItem
          issue={issue}
          onToggleContent={action('toggle')}
          onToggleLock={action('toggle lock')}
        />
      </PreviewArea>
      <Code>
        {`
          <IssueItem issue={issue} onToggleContent={action('toggle')} onToggleLock={action('toggle lock')} />
        `}
      </Code>
      {propsDetail}
    </>
  ),
);
