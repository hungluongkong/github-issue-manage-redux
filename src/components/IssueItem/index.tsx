import React from 'react';
import styled from 'styled-components';
import Button from '../Button';

import * as color from '../../theme/color';
import * as metrics from '../../theme/metrics';
import * as localize from '../../constants/localize';
import { isEqualAllProps } from '../../helpers/propsHelper';
import {
  IssueProps,
  ButtonType,
} from '../../types';

// Main container
const IssueWrapper = styled.div<{isLocked: boolean}>`
  display: flex;
  box-sizing: border-box;
  padding: 5px;
  justify-content: space-between;
  background: ${(props) => (props.isLocked ? color.ISSUE_BG_LOCKED : color.ISSUE_BG_DEFAULT)};

  :hover {
    opacity: ${metrics.OPACITY.hover};
    box-shadow: 2px 1px 2px 1px ${color.SHADOW_BORDER};
  }

  :before {
    content: ${(props) => (props.isLocked ? '"ðŸ”’"' : '"ðŸ”“"')};
    padding: 7px 0 10px 5px;
    font-size: ${metrics.FontSize.Icon};
  }
`;

// Display Id in github of issue. On left - middle
const IssueID = styled.div`
  font-weight: 600;
  text-align: center;
  padding: 10px 20px 10px 0px;
  min-width: 30px;
`;

// Wrap title & date. Export for test
export const TitleWrapper = styled.div`
  flex-grow: 1;
  :hover {
    cursor: pointer;
  }
`;

// Big issue title
export const IssueTitle = styled.a`
  font-weight: 700;
  display: block;
`;

// Edited/create date under the title.
const EditTime = styled.span`
  display: block;
  color: ${color.SMALL_TEXT_COLOR};
  font-size: ${metrics.FontSize.Small};
  margin-top: 4px;
  line-height: 1.25;
`;

const IssueItem = ({
  issue,
  onToggleContent,
  onToggleLock,
}: IssueProps): JSX.Element => (
  <IssueWrapper isLocked={issue.locked}>
    {/* Issue number is id of issue in user's storage list */}
    <IssueID>#{issue.number}</IssueID>

    {/* Body of issue item */}
    <TitleWrapper onClick={() => { onToggleContent(issue); }}>
      <IssueTitle>{issue.title}</IssueTitle>
      <EditTime>
        {
          // Compare edit/create timestamp
          new Date(issue.createdAt).getTime() === new Date(issue.updatedAt).getTime()
            ? localize.en.ISSUE_CREATED_AT
            : localize.en.ISSUE_UPDATED_AT
        }
        { new Date(issue.updatedAt).toLocaleDateString() }
      </EditTime>
    </TitleWrapper>
    {
      <Button
        type={issue.locked ? ButtonType.Primary : ButtonType.Default}
        onClick={() => { onToggleLock(issue); }}
        value={issue.locked ? localize.en.BTN_ISSUE_UNLOCK : localize.en.BTN_ISSUE_LOCK}
      />
    }
  </IssueWrapper>
);

export default React.memo(IssueItem, isEqualAllProps);
