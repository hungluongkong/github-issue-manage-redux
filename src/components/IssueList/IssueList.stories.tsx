import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import IssueList from '.';

import { JsonToIssueList } from '../../helpers/jsonHelper';
import { issueListJson } from '../../constants/storiesTempData';
import { PreviewArea, Code, DetailTable } from '../commonStyled';

// Detail of props
const propsDetail: JSX.Element = DetailTable({
  data: [
    {
      name: 'issues',
      type: 'Array of Issue',
      description: 'Issue list to display',
    },
    { name: 'loading', type: 'boolean', description: 'is List still loading' },
  ],
});

const issues = JsonToIssueList(issueListJson);

storiesOf('Issue List', module)
  .add(
    'Default',
    (): JSX.Element => (
      <>
        <h3>Default List</h3>
        <PreviewArea>
          <IssueList
            issues={issues}
            isShowForm={false}
            task=""
            lastTaskTime={new Date()}
            toggleForm={action('toggle Form')}
            getIssues={action('get')}
            toggleIssue={action('toggle Issue')}
            toggleLockIssue={action('lock/unlock an issue')}
          />
        </PreviewArea>
        <Code>
          {`
          <IssueList
            issues={issues}
            isShowForm={false}
            task=""
            lastTaskTime={new Date()}
            toggleForm={action('toggle')}
            getIssues={action('get')}
            toggleLockIssue={action('lock/unlock an issue')}
          />
        `}
        </Code>
        {propsDetail}
      </>
    ),
  )
  .add(
    'Loading',
    (): JSX.Element => (
      <>
        <h3>Loading List</h3>
        <PreviewArea>
          <IssueList
            loading
            issues={issues}
            isShowForm={false}
            task=""
            lastTaskTime={new Date()}
            getIssues={action('get')}
            toggleForm={action('toggle')}
            toggleIssue={action('toggle Issue')}
            toggleLockIssue={action('lock/unlock an issue')}
          />
        </PreviewArea>
        <Code>
          {`
          <IssueList
            loading
            issues={issues}
            isShowForm={false}
            task=""
            lastTaskTime={new Date()}
            getIssues={action('get')}
            toggleForm={action('toggle')}
            toggleLockIssue={action('lock/unlock an issue')}
          />
        `}
        </Code>
        {propsDetail}
      </>
    ),
  );
