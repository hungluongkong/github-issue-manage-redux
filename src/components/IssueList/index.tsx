/* eslint-disable no-nested-ternary */
import React, { useEffect } from 'react';
import styled from 'styled-components';
import IssueItem from '../IssueItem';
import LoadingBar from '../LoadingBar';
import ErrorBoundary from '../ErrorBoundary';
import Button from '../Button';
import TaskReporter from '../TaskReporter';

import { isEqualAllProps } from '../../helpers/propsHelper';

import * as color from '../../theme/color';
import * as metrics from '../../theme/metrics';
import * as localize from '../../constants/localize';
import {
  Issue,
  IssueListProps,
  ButtonType,
} from '../../types';

// Wrap all content
const IssueListWrapper = styled.div`
  background: ${color.ISSUE_LIST_BG};
`;

// Heading style
const ListHeading = styled.div`
  background: ${color.ISSUE_LIST_HEAD_BG};
  text-align: center;
  font-size: ${metrics.FontSize.Title};
`;

const IssueList = ({
  lastTaskTime,
  isShowForm,
  getIssues,
  toggleForm,
  toggleIssue,
  toggleLockIssue,
  loading = false,
  issues = [],
  task = '',
}: IssueListProps): JSX.Element => {
  // Fetch issues on start
  useEffect(() => {
    if (issues.length === 0) {
      getIssues();
    }
    // eslint-disable-next-line
  }, []);

  return (
    <ErrorBoundary>
      <TaskReporter
        content={task}
        timeoutMS={metrics.POPUP_TIMEOUT}
        date={lastTaskTime}
      />
      <Button
        type={ButtonType.Primary}
        value={isShowForm ? localize.en.BTN_FORM_CLOSE : localize.en.BTN_ISSUE_ADD}
        onClick={() => {
          toggleForm();
        }}
      />
      <ListHeading>Issues</ListHeading>
      <IssueListWrapper>
        {
          // If loading or Empty list
          loading || issues.length === 0 ?
            (
              loading ? <LoadingBar /> : localize.en.ISSUE_LIST_NO_ISSUE
            )
            :
            // List render
            (
              issues.map((issue: Issue): JSX.Element => (
                <IssueItem
                  key={issue.id}
                  issue={issue}
                  onToggleContent={toggleIssue}
                  onToggleLock={toggleLockIssue}
                />
              ))
            )
        }
      </IssueListWrapper>
    </ErrorBoundary>);
};

export default React.memo(IssueList, isEqualAllProps);
