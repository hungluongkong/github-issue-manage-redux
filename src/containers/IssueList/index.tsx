import { connect } from 'react-redux';
import { RootReducerState } from '../../types';
import IssueList from '../../components/IssueList';
import { getIssues, updateIssue, toggleLockIssue } from '../../actions/issues';
import { toggleIssue, toggleForm } from '../../actions/asideView';

const mapStateToProps = (state: RootReducerState) => ({
  issues: state.issues.issues,
  loading: state.issues.pending,
  task: state.issues.task,
  lastTaskTime: state.issues.lastTaskTime,
  isShowForm: state.asideView.isShowForm,
});

const mapDispatchToProps = {
  getIssues,
  updateIssue,
  toggleIssue,
  toggleForm,
  toggleLockIssue,
};

export default connect(mapStateToProps, mapDispatchToProps)(IssueList);
