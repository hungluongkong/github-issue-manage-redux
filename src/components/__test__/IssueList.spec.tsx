import React from 'react';
import { shallow } from 'enzyme';
import { create } from 'react-test-renderer';

import IssueList from '../IssueList';
import IssueItem from '../IssueItem';
import LoadingBar from '../LoadingBar';

import { IssueListProps } from '../../types';
import { JsonToIssueList } from '../../helpers/jsonHelper';
import { issueListJson } from '../../constants/storiesTempData';

describe('Unit test for IssueList component', () => {
  const issues = JsonToIssueList(issueListJson);
  const defaultProps: IssueListProps = {
    issues,
    getIssues: jest.fn(),
    toggleForm: jest.fn(),
    toggleIssue: jest.fn(),
    toggleLockIssue: jest.fn(),
    lastTaskTime: new Date(),
    loading: false,
    task: '',
    isShowForm: false,
  };

  it('snapshot', () => {
    const issueList = create(<IssueList {...defaultProps} />);
    expect(issueList).toMatchSnapshot();
  });

  it('should display all issue', () => {
    const issueList = shallow(<IssueList {...defaultProps} />);
    expect(issueList.find(IssueItem).length).toEqual(issues.length);
  });

  it('should display loading', () => {
    const issueList = shallow(
      <IssueList {...{ ...defaultProps, loading: true }} />,
    );
    expect(issueList.find(LoadingBar).exists()).toEqual(true);
  });
});
