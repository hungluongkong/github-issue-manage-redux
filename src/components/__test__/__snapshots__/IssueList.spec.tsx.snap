// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for IssueList component snapshot 1`] = `
Array [
  <div
    className="sc-fzozJi jRiTbz"
    style={
      Object {
        "visibility": "hidden",
      }
    }
    type={0}
  >
    <p
      className="sc-fzoLsD fEbxod"
    >
      
    </p>
  </div>,
  <button
    className="sc-AxjAm caeeVY"
    onClick={[Function]}
    size={1}
  >
    New Issue
  </button>,
  <div
    className="sc-fzplWN vdWWO"
  >
    Issues
  </div>,
  <div
    className="sc-fzpans hbgWpR"
  >
    <div
      className="sc-AxirZ bggFi"
    >
      <div
        className="sc-AxiKw cEQHGP"
      >
        #
        1
      </div>
      <div
        className="sc-AxhCb bREgZd"
        onClick={[Function]}
      >
        <a
          className="sc-AxhUy ogOxz"
        >
          isssue no 1
        </a>
        <span
          className="sc-AxgMl gaiwNa"
        >
          Created at 
          7/9/2020
        </span>
      </div>
      <button
        className="sc-AxjAm gZMYpY"
        onClick={[Function]}
        size={1}
      >
        Lock Issue
      </button>
    </div>
    <div
      className="sc-AxirZ bggFi"
    >
      <div
        className="sc-AxiKw cEQHGP"
      >
        #
        2
      </div>
      <div
        className="sc-AxhCb bREgZd"
        onClick={[Function]}
      >
        <a
          className="sc-AxhUy ogOxz"
        >
          #practice #make #api
        </a>
        <span
          className="sc-AxgMl gaiwNa"
        >
          Created at 
          7/10/2020
        </span>
      </div>
      <button
        className="sc-AxjAm gZMYpY"
        onClick={[Function]}
        size={1}
      >
        Lock Issue
      </button>
    </div>
  </div>,
]
`;
